### PR TITLE
DFA fix from LRydin's fork

### PR DESCRIPTION
**Pull request by Tam-Pham**
As discussed in https://github.com/neuropsychology/NeuroKit/issues/206, @LRydin has identified an issue with the implementation of MFDFA and has started the fix for it. 

As we are urgently in need to use this function for our upcoming analysis, I have *blatantly* made this PR on @LRydin's behalf 😅 to continue working on it. Hope it's alright with you and please do take a look whenever you have some time! 😀 

# Checklist

Here are some things to check before creating the PR. If you encounter any issues, do let us know :)

- [x] I have read the [CONTRIBUTING](https://github.com/neuropsychology/NeuroKit/blob/master/.github/CONTRIBUTING.rst#structure-and-code) file.
- [x] My PR is targetted at the **dev branch** (and not towards the master branch).
- [x] I ran the [CODE CHECKS](https://github.com/neuropsychology/NeuroKit/blob/master/.github/CONTRIBUTING.rst#run-code-checks) on the files I added or modified and fixed the errors.
- [ ] I have added the newly added features to **News.rst** (if applicable)

### CODE REVIEW

File: neurokit2/hrv/hrv_nonlinear.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 from .hrv_utils import _hrv_get_rri, _hrv_sanitize_input
 
 
-def hrv_nonlinear(peaks, sampling_rate=1000, show=False, **kwargs):
+def hrv_nonlinear(peaks, sampling_rate=1000, dfa_short_window=[4, 16], dfa_long_window=[16, 64], show=False, **kwargs):
```
Comment (zen-juen) [inline neurokit2/hrv/hrv_nonlinear.py:16]: not sure if there is a better way of naming these arguments 😄

**Comment by LRydin** [comment]
@Tam-Pham This is good, it puts some pressure on me to get things done :P which I should. I will work on this *right now* to hopefully have an acceptable state by the end of the day. Sorry for all the waiting... I always say I'll get it done... and then time flies by...

**Comment by LRydin** [comment]
I began restructuring [`fractal_dfa()`](https://github.com/neuropsychology/NeuroKit/blob/master/neurokit2/complexity/fractal_dfa.py) in this [commit](https://github.com/LRydin/NeuroKit/commit/ff0fbc1e0600497bd92a50bf3df9649ef5d03243). There is a clear issue with the way `windows` is being used in the MFDFA part.  I'm on it now.

File: neurokit2/hrv/hrv_nonlinear.py (hunk 1 above)
Comment (DominiqueMakowski) [reply neurokit2/hrv/hrv_nonlinear.py:16]: `dfa_windows = [[4, 16], [16, 64]]`? Not the best to have nested lists but... It's extendable and parsimonious in terms of arguments number

**Comment by LRydin** [comment]
Alright, a further step. I am introducing a new module (this [commit](https://github.com/LRydin/NeuroKit/commit/dac6ebf2a34bfac8986b9c0f47bc6eef14a9fa46)), denoted in my fork as [fractal_measures](https://github.com/LRydin/NeuroKit/blob/master/neurokit2/complexity/fractal_dfa.py).

@DominiqueMakowski @Tam-Pham @zen-juen, I would love your comments on this, if possible. Below the gist, thus far:

----

Imports
```python
from neurokit2.complexity import singularity_spectrum, scaling_exponents, hurst_exponents
from scipy.stats import levy_stable   # needed for making mock timeseries
import matplotlib.pyplot as plt
```
Let us start by generating a multifractal time series (actuall a bi-fractal timeseries, but is goes to show)

```python
X = levy_stable.rvs(alpha=1.5, beta = 0, size=10000)
```
now we will extract the [*singularity spectrum*](https://en.wikipedia.org/wiki/Singularity_spectrum), which on the background needs to run `fractal_mfdfa()` (which is just `fractal_dfa()` with `multifractal = True` and a set of `q` powers).

```python
alpha, f = singularity_spectrum(X)
plt.plot(alpha, f)
```
Screenshot from 2021-08-13 17-01-08

Now we get the *singularity spectrum* (no wiki on this)

```python
q, tau = scaling_exponents(X)
plt.plot(q, tau)
```
multifractal scaling exponents

lastly, the _generalised Hurst coefficients_

```python
q, hq = hurst_exponents(X)
plt.plot(q, hq)
```
Generalised Hurst coefficients

----

These plots are always hard to interpret, but there are some known measures we can immediately draw out of this. For example, the *singularity spectrum* usually looks like this:

download

and you can see that in the picture they identify two measures: the vertical height and horizontal width. They carry a meaning in quantifying the "fractality" of a time series.
I don't know if you use some sort of "report function" in NK, but I could imagine a simple "multifractal_measures" functions that outputs Δα and Δf(α) for the user.

How do you see this :)?

**Comment by Tam-Pham** [comment]
Hey @LRydin, 

I feel bad putting the pressure on you but glad you found it alright ^^

I think you re-structuring of [fracta_dfa()](https://github.com/LRydin/NeuroKit/blob/master/neurokit2/complexity/fractal_dfa.py) is great. The few styling mismatching with the current style of neurokit can be fixed later. 

However, I do have a few thoughts about the [fractal_measure()](https://github.com/LRydin/NeuroKit/blob/master/neurokit2/complexity/fractal_measures.py):

Basically, for the singularity spectrum, from my understanding, we will need to first calculate `Hq` and then `tau` and then the two singularity indices `hq` and `Dq`. I'm thinking for this specific file, can we keep it shorter, especially for neurokit, I feel like the recalculating of windows and fluctuations in a few main functions might not be necessary. Hence, I propose that we can combine the internal functions with their respective main functions. Specifically, below are my few suggestions:

- To calculate the `Hurst exponent (Hq)` for the range of q values. I wonder if:
   - (1) for the [_slope()](https://github.com/LRydin/NeuroKit/blob/85473622beecc5a02c6e55f56b38819f75aa3a98/neurokit2/complexity/fractal_measures.py#L561), can we keep it standardized with the script in [fractal_measure](https://github.com/LRydin/NeuroKit/blob/85473622beecc5a02c6e55f56b38819f75aa3a98/neurokit2/complexity/fractal_dfa.py#L124):

```python
dfa = np.zeros(len(q))
for i in range(len(q)):
    dfa[i] = np.polyfit(np.log2(windows), np.log2(fluctuations[:,I]), 1)[0]
```
   - (2) instead of having a `_slope()` function and two `hurst_exponents()` functions, can we combine all of them into one main `hurst_exponents()` function. So something like:
 ```python
def hurst_exponents(windows, fluctuations, q):
    Hq = np.zero(len(q))
    for i in range(len(q)):
        Hq[i] = np.polyfit(np.log2(windows), np.log2(fluctuations[:,I]), 1)[0]
    return Hq
```

- To calculate the mass scaling exponent or tau, I wonder if we can also have just one function `scaling_exponent()` where `tau(q) = qh(q) - 1` is calculated.

- And last but not least, I think your implementation of the singularity exponent (`alpha` or `hq`) and singluarity dimension (`falpha` or `Dq`) is perfect 👍 Haha I have nothing to comment. 

I do very much like the idea of plotting `tau` and `Hq` for the the range of q values. I propose that we can create a common plotting function to plot all of these plots together using `pyplot.subplots`. When `show=True`, we can return all the plots together. Anyway, please don't worry about this first. I can help to look into this once you make a PR into this branch.

- Regarding:
> I don't know if you use some sort of "report function" in NK, but I could imagine a simple "multifractal_measures" functions that outputs Δα and Δf(α) for the user.

This is quite an API matter and I'm too not sure what's the best way to return these singularity spectrum indices.  Personally, I feel like we can combine the fractal_dfa and fractal_measures files together and return just the slope for when multifractal is False and return a dictionary contains Δα and Δf(α) (and other indices of singularity spectrum) when multifractal is True. What do you guys think @DominiqueMakowski @zen-juen 

@LRydin do let me know what you think and if I miss anything in reviewing your changes. If you happen to have no time on your hands, do not hesitate to make the PR from your fork to this branch. I can make the changes for your review 😄

**Comment by LRydin** [comment]
@Tam-Pham absolutely zero worries for putting me on the spot. I said I wanted to contributed on NK2 back in the dark ages and I never got around to it - so it was most welcomed, your little nudge :).

Initially I did start the whole this as a single module, but then I thought it would get immeasurably long. I am happy to combine everything into one module. 

I'll try to recombine things into single functions. I got into the habit of using `_sneaky_functions_after_an_underscore()` to do the heavy lifting, but that is just me :P. I'll standardised to your solution for `_slopes()` and so on :).

I'll get it done tomorrow :)!

**Comment by Tam-Pham** [comment]
P/S: do add yourself to the [list of contributors](https://github.com/neuropsychology/NeuroKit/blob/7c34091cfdce2b47f3b2a312bd59c8c93d921bcf/AUTHORS.rst)

File: neurokit2/hrv/hrv_nonlinear.py [hunk 2]
```diff
@@ -5,15 +5,16 @@
...
+                          entropy_sample)
 from ..complexity.fractal_correlation import fractal_correlation
+from ..complexity.fractal_dfa import fractal_dfa
 from ..misc import find_consecutive
 from ..signal import signal_zerocrossings
 from .hrv_utils import _hrv_get_rri, _hrv_sanitize_input
 
 
-def hrv_nonlinear(peaks, sampling_rate=1000, show=False, **kwargs):
+def hrv_nonlinear(peaks, sampling_rate=1000, dfa_windows=[(4, 11), (12, None)], show=False, **kwargs):
```
Comment (DominiqueMakowski) [inline neurokit2/hrv/hrv_nonlinear.py:17]: i don't think we should expose this argument here, but allow it through kwargs

File: neurokit2/hrv/hrv_nonlinear.py [hunk 3]
```diff
@@ -124,10 +131,37 @@ def hrv_nonlinear(peaks, sampling_rate=1000, show=False, **kwargs):
 
             - **RCMSE**: The refined composite multiscale entropy measure of HRV, calculated by `entropy_multiscale()`.
 
-            - **CorrDim**: The correlation dimension of the HR signal, calculated by `fractal_correlation()`.
+            - **CD**: The correlation dimension of the HR signal, calculated by `fractal_correlation()`.
```
Comment (DominiqueMakowski) [inline neurokit2/hrv/hrv_nonlinear.py:134]:
```suggestion
            - **CD**: The Correlation Dimension of the HR signal, calculated by `fractal_correlation()`.
```

File: neurokit2/hrv/hrv_nonlinear.py [hunk 4]
```diff
@@ -387,6 +420,56 @@ def _hrv_nonlinear_fragmentation(rri, out):
...
+    long_window = np.linspace(dfa_windows[1][0], int(max_beats), n_windows_long).astype(int)
+
+    # For monofractal
+    out["DFA_alpha1"] = fractal_dfa(rri, multifractal=False, windows=short_window, **kwargs)['slopes'][0]
+    out["DFA_alpha2"] = fractal_dfa(rri, multifractal=False, windows=long_window, **kwargs)['slopes'][0]
+
+    # For multifractal
+    mdfa_alpha1 = fractal_dfa(rri,
+                              multifractal=True,
+                              q=[-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5],
```
Comment (DominiqueMakowski) [inline neurokit2/hrv/hrv_nonlinear.py:456]:
```suggestion
                              q=range(-5, 5),
```

(note that, in `fractal_dfa()` u need to convert the `range` object to a list

File: tests/tests_hrv.py [hunk 5]
```diff
@@ -73,7 +73,7 @@ def test_hrv():
                'HRV_C2a', 'HRV_SD2d', 'HRV_SD2a',
                'HRV_Cd', 'HRV_Ca', 'HRV_SDNNd',
                'HRV_SDNNa', 'HRV_ApEn', 'HRV_SampEn', 'HRV_MSE', 'HRV_CMSE',
-               'HRV_RCMSE', 'HRV_DFA', 'HRV_CorrDim']
+               'HRV_RCMSE', 'HRV_DFA_α1', 'HRV_DFA_α2', 'HRV_CD']
```
Comment (DominiqueMakowski) [inline tests/tests_hrv.py:76]: non-ascii = 🙅‍♂️

File: neurokit2/hrv/hrv_nonlinear.py (hunk 4 above)
Comment (Tam-Pham) [reply neurokit2/hrv/hrv_nonlinear.py]: I used np.arange(-5, 6) instead

**Comment by Tam-Pham** [comment]
btw, @LRydin, after re-organizing and tidying the code, everything seems to work well but do not hesitate to give it a look when you have time. I also combined the plotting functions into one. So for a sample signal like below:

```python
import neurokit2 as nk
import numpy as np

signal = nk.signal_simulate(duration=10, noise=0.05)
dfa = nk.fractal_mfdfa(signal, q=np.arange(-5, 6), show=True)
```
A combined plots will be returned:
image

Thanks much for your help!

**Pull request merged**